packages/ariakit/src/menu/menu-item.ts

```diff
@@ -37,13 +42,14 @@ export const useMenuItem = createHook<MenuItemOptions>(
     state = useStore(state || context, ["move"]);
 
     const onClickProp = useEventCallback(props.onClick);
+    const hideOnClickProp = useBooleanEventCallback(hideOnClick);
     const hideMenu = state && "hideAll" in state ? state.hideAll : undefined;
 
     const onClick = useCallback(
       (event: MouseEvent<HTMLDivElement>) => {
         onClickProp(event);
         if (event.defaultPrevented) return;
-        if (!hideOnClick) return;
+        if (!hideOnClickProp(event)) return;
         if (!hideMenu) return;
         // If this item is also a menu button, we don't want to hide the menu.
         const popupType = event.currentTarget.getAttribute("aria-haspopup");
```

Comment:
Since it's not just a boolean prop anymore, we should have this `hideOnClickProp(event)` check as the last check before calling `hideMenu`. This is because users could use `hideOnClick` as a function to run some side effects, and we don't want to trigger these side effects on click events that wouldn't have called hideMenu in the first place.

Reply:
That makes a lot of sense :+1: